### ISSUE TITLE
Give NVDA the ability to detect time until full charge.

### ISSUE THREAD

**Issue by derekriemer** [nvaccess/nvda#6309]
In windows ten, it is the case that windows can tell you how many minutes until a full charge. I'm not sure how, because https://msdn.microsoft.com/en-us/library/windows/desktop/aa373232(v=vs.85).aspx doesn't have this built in, and neither does another windows api. I might play with this and see if there's a way to bring this feature in to reportBatteryStatus, so it reports something like "On AC power, 20min until full charge".
##STR:
0. Discharge your computer to about 50%.
1. Plug it in.
1. press windows+b
2. go to the battery status.

It should say something like "20 minutes until full charge" after some time being plugged in.

**Comment by nromey** [comment]
I'll have to look at the code for getting the battery status.  I notice though that when you look in the system tree manually, it shows this information.  There has to be a way. and agree, if the feature is not currently in NVDA, I think that this would be a useful one, and maybe, it's not something that would take a whole lot of work to make happen if the information is available somewhere, even if we access it via tixt in the system tray battery information icon.

**Comment by derekriemer** [comment]
I don't see it with system Power Info, but windows is getting it from somewhere. It could be calculating it manually.

**Comment by nromey** [comment]
I don't see anything in that function method.  It looks like Windows, and Mac OSX for that matter, go through a calculating phase where they determine rate of charge based on current etc. and teh power acpi interface.

What you could do, and I'm not sure if I could make this happen every time would be to determine if (a) windows 10 was running.  Then (b) if we're in windows 10  and AC was connected, the script does this already,then it would have to do the following.

Grab the button object in the system tray where it displays battery information.  It would be clunky but you could grab time to charge from that button/system tray battery icon.

**Comment by josephsl** [comment]
Hi, if it can be found, I think an API-based method would be a bit better. Thanks.

**Comment by ruifontes** [comment]
The most strange thing is that sometimes the information appears in the 
system tray button and sometimes appears only the capacity percentage...

In this moment, in the systray appears the time remaining to full charge, 1 
hour 20 minutes and in Insert+Shift+B is announced the capacity 
percentage...
After writing the above sentence I was checking and in Systray appears only 
the percentage...

Rui





-----Mensagem Original----- 
De: nromey
Data: 27 de agosto de 2016 01:16
Para: nvaccess/nvda
Assunto: Re: [nvaccess/nvda] Give NVDA the ability to detect time until full 
charge. (#6309)



I don't see anything in that function method. It looks like Windows, and Mac 
OSX for that matter, go through a calculating phase where they determine 
rate of charge based on current etc. and teh power acpi interface.

What you could do, and I'm not sure if I could make this happen every time 
would be to determine if (a) windows 10 was running. Then (b) if we're in 
windows 10 and AC was connected, the script does this already,then it would 
have to do the following.

Grab the button object in the system tray where it displays battery 
information. It would be clunky but you could grab time to charge from that 
button/system tray battery icon.

-

**Comment by derekriemer** [comment]
Could this be done reliably fo multiple languages? 

Sent from a mobile device.
Please disregard errors as this is a smaller device.

**Comment by leonardder** [comment]
I've looked into this twice, and twice stumbled upon impossibilities. My findings:

1. The current approach with [GetSystemPowerStatus](https://msdn.microsoft.com/nl-nl/library/windows/desktop/aa373232(v=vs.85).aspx) can't retrieve time until full charge. There is one interesting parameter though, reserved1. That one now returns whether battery saver is on or off. So showing that could be implemented.
2. The approach using the [Win32_Battery WMI class](https://msdn.microsoft.com/en-us/library/aa394074(v=vs.85).aspx).

	Time required to fully charge the battery. This property is not supported. BatteryRechargeTime does not have a replacement property, and is now considered obsolete.

I'd like to vote against getting information from the system tray, as this is not at all reliable and quite error-prone, since the icon text will vary between languages.

**Comment by Brian1Gaff** [comment]
In my view this is going to be some piece of bespoke software in any given 
laptop, but from what people tell me this is not a perfect art. its a bit 
like file copy and stuff like that  where the  time estimated can seriously 
change over time.
 I have a washing machine that works like this with time to end of its 
cycle!
 Normally you cannot overcharge well designed laptops, so is this really 
needed?
 The estimated time till the battery warning is after all displayed.
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal email to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
----- Original Message -----

**Comment by nromey** [comment]
I agree w the no system tray option. I found that if today items are rearranged, the approach fails. I think Windows probably samples battery percentages. Could one poll percentages in the background with the kernel method and mathematically calculate it. I'm not sure but it most likely wouldn't take too much processor work. I'm just not sure what the project policies are about events like that.

**Comment by derekriemer** [comment]
[**@leonardder**](https://github.com/leonardder) commented on [Aug 27, 2016, 1:57 AM MDT](https://github.com/nvaccess/nvda/issues/6309#issuecomment-242903250 "2016-08-27T07:57:22Z - Replied by Github Reply Comments"):
> I've looked into this twice, and twice stumbled upon impossibilities. My findings:
> 1. The current approach with [GetSystemPowerStatus](https://msdn.microsoft.com/nl-nl/library/windows/desktop/aa373232%28v=vs.85%29.aspx) can't retrieve time until full charge. There is one interesting parameter though, reserved1. That one now returns whether battery saver is on or off. So showing that could be implemented.
> 2. The approach using the [Win32_Battery WMI class](https://msdn.microsoft.com/en-us/library/aa394074%28v=vs.85%29.aspx).
> 
> > Time required to fully charge the battery. This property is not supported. BatteryRechargeTime does not have a replacement property, and is now considered obsolete.
> 
> I'd like to vote against getting information from the system tray, as this is not at all reliable and quite error-prone, since the icon text will vary between languages.

Can you describe [2]

**Comment by leonardder** [comment]
I played somewhat with the answer in
http://stackoverflow.com/questions/21083518/get-battery-status-using-wmi-in-python
    This uses win32com though, thus you have to run from source or in a
    stand alone interpreter to test.

**Comment by Adriani90** [comment]
Is this also valid for Windows 11 24 H2? Can sighted people see this info in the system tray?
cc: @Qchristensen, @seanbudd maybe you can help here with your eyes? ;-)

**Comment by seanbudd** [comment]
@Adriani90 - No I don't think there is a way to get this info from the system tray. I think the only way to get this as a user is from `Settings > System > Power &Battery`.  I think this can be fetched automatically with [TimeToFullCharge](https://learn.microsoft.com/en-us/windows/win32/cimwin32prov/win32-battery?redirectedfrom=MSDN), and shouldn't be too much effort to add to existing power status reporting.

**Comment by nidza07** [comment]
@seanbudd My system tray icon says the following when the battery is charging:
"Battery status: 15 min to full charge"

I think this could well be hidden on some laptops though, because I definitely had devices where this information is unavailable.
